### PR TITLE
Spack content

### PR DESCRIPTION
**Pull request by jhodrien**
Addresses #3 
Review and comments welcome

### CODE REVIEW

File: book/course/spack/testinstall.md [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+  Fetch: 0.67s.  Build: 2.23s.  Total: 2.89s.
+[+] /home/me/spack/opt/spack/linux-rhel8-haswell/gcc-8.5.0/pigz-2.7-qkx6pxv5qtwcjsmmfrqzbo32g5uwkshy
+```
+
+That is quite wordy, but to note, it's installed pigz, and also installed a
+dependency of it, zlib.  Just asking for it to install that one package has led
+to it downloading and installing another, and it's quite happy doing that
+without needing any guidance.
+
+Let's now see how you can load that module, and use it.  If you're using
```
Comment (Sparrow0hawk) [inline book/course/spack/testinstall.md:36]: Potentially, need to specifically call out our HPC audience here, who are using environment modules?

File: book/course/spack/modules.md [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+In the last section we built a version of pigz, and loaded it:
+
+```bash
+$ module add pigz-2.7-gcc-8.5.0-qkx6pxv
+$ pigz --version
+pigz 2.7
+```
+
+Now, it's very definitive what you're loading (pigz version 2.7 built with gcc
+8.5.0 with a hash to pin it down further), but it's a bit wordy.  This format
```
Comment (Sparrow0hawk) [inline book/course/spack/modules.md:12]: Are we assuming the audience understand what a `hash` is here? Do we need to know there's a hash or could we just say it's got a big long unwieldy module name?

Not sure of the right answer just trying to balance details v. cognitive load for the audience

Comment (Sparrow0hawk) [reply book/course/spack/modules.md:12]: You sort of touch on this later on, so inclined to leave it as is

File: book/course/spack/existing.md [hunk 3]
```diff
@@ -0,0 +1,244 @@
+# Using existing software
+
+Spack by default will quite happily just build everything itself, as long as a
+limited number of [system requirements](https://spack.readthedocs.io/en/latest/getting_started.html#system-prerequisites)
+are satisfied.  But you might want to use a compiler and MPI library from a
+host system.  This is often the case if packages take a long time to build, or
+if you've got access to optimised versions of software that you don't want to
+replicate yourself.
```
Comment (Sparrow0hawk) [inline book/course/spack/existing.md:8]:
```suggestion
replicate yourself such as on ARC3 and ARC4.
```
Or to that effect.

File: book/course/spack/existing.md [hunk 4]
```diff
@@ -0,0 +1,244 @@
...
+Spack about your host system, and the installed software is up to you.  On one
+hand you can massively reduce how much software you need to build yourself, but
+on the other hand, you become much more dependent on the host system.  There's
+not one right answer.
+
+## Compilers
+
+Compilers are a bit different to other pieces of software.  Typically the
+easiest way to get Spack to use a system compiler is to make sure it's in your
+path, and then tell Spack to find it:
```
Comment (Sparrow0hawk) [inline book/course/spack/existing.md:21]:
```suggestion
path, and tell Spack to find it:
```

File: book/course/spack/existing.md [hunk 5]
```diff
@@ -0,0 +1,244 @@
...
+
+```bash
```
Comment (Sparrow0hawk) [inline book/course/spack/existing.md:23]: I'm confused by the below code snippet. You have this compiler already loaded as a module? Or is it just the default compiler in your PATH?

Maybe adding a `module load` call or `gcc -v`  before the `spack compiler add` would just help clarify that?

Comment (Sparrow0hawk) [reply book/course/spack/existing.md:23]: You address this slightly below (line 31/32) but I think just for the snippets sake it might be worth adding just `gcc -v` or some text clarifying you've got gnu/8.5.0 installed as the default

File: book/course/spack/existing.md [hunk 6]
```diff
@@ -0,0 +1,244 @@
...
+spack config edit packages
+```
+
+And tell it to use this compiler
+
+```yaml
+packages:
+  all:
+    compiler:
+    - 'gcc@8.3.0' # Your preferred compiler here
```
Comment (Sparrow0hawk) [inline book/course/spack/existing.md:48]: Just a sanity check, you use 8.5.0 above but have 8.3.0 here?

File: book/course/spack/existing.md [hunk 7]
```diff
@@ -0,0 +1,244 @@
...
+  libjpeg-turbo:
+    externals:
+    - spec: libjpeg-turbo@1.5.3
+      prefix: /usr
+```
+
+There's a whole lot of interesting stuff in there that I've not had to think
+about, as Spack has added all the details about the openmpi variant it
+discovered.  We'll cover some of this in the advanced building section.
+Annoyingly, it got one detail wrong because of how Red Hat split the
```
Comment (Sparrow0hawk) [inline book/course/spack/existing.md:175]: This might warrant an explicit admonition-style callout as it's more a gotcha-hint than general tutorial material?

File: book/course/spack/advanced.md [hunk 8]
```diff
@@ -0,0 +1,194 @@
...
+Now Spack will happily rebuild that specific variant you've asked for, with the
+compiler you've asked for.
+
+## Dependencies
+
+You can also specify particular versions and variants of dependencies.  We can
+take our previous build even further:
+
+```bash
+spack install piranha @develop +ipo -python build_type=Release ^boost@1.62.0+thread %gcc@11.2.1
```
Comment (Sparrow0hawk) [inline book/course/spack/advanced.md:88]: Can you clarify the distinction of using the `^` for boost but the `%` for gcc in the text?
My assumption is package v compiler but might be worth clarifying.

**Review by Sparrow0hawk — CHANGES_REQUESTED** [review]
I think this is really excellent, you've gone above and beyond what I think we'd want for the course but it's perfect to include that as extra reading.

I'd be interested to know which bits you'd want to cover in a taught session. My hunch would be:
- Install
- Test install
- Modules
- Existing 

You've gone for an alternate structure to me (multiple subpages) but I think at this stage we keep it like this until we've got some more sections before editorialising it as a whole.

File: book/course/spack/modules.md (hunk 2 above)
Comment (ollie-clark) [reply book/course/spack/modules.md:12]: I think it's still worth giving a basic explanation of what a hash is (a unique identifier for this particular version of the software).

**Review by Sparrow0hawk — APPROVED** [review]
Looks great, merge away
